code/modules/gamemaster/event2/event.dm

```diff
@@ -0,0 +1,230 @@
+	var/start_delay = null
+	var/announce_delay = null
+	var/length = null
+
+// Returns the name of where the event is taking place.
+// In the future this might be handy for off-station events.
+/datum/event2/event/proc/location_name()
+	return station_name()
+
+// Returns the z-levels that are involved with the event.
+// In the future this might be handy for off-station events.
+/datum/event2/event/proc/get_location_z_levels()
+	return using_map.station_levels
+
+// Returns a list of empty turfs in the same area.
+/datum/event2/event/proc/find_random_turfs(minimum_free_space = 5, list/specific_areas = list(), ignore_occupancy = FALSE)
+	var/list/area/grand_list_of_areas = find_random_areas(specific_areas)
+
+	if(!LAZYLEN(grand_list_of_areas))
+		return list()
+
+	for(var/thing in grand_list_of_areas)
+		var/list/A = thing
+		var/list/turfs = list()
+		for(var/turf/T in A)
+			if(!T.check_density())
+				turfs += T
+
+		if(turfs.len < minimum_free_space)
+			continue // Not enough free space.
+		return turfs
+
+	return list()
+
+/datum/event2/event/proc/find_random_areas(list/specific_areas = list(), ignore_occupancy = FALSE)
+	if(!LAZYLEN(specific_areas))
+		specific_areas = global.the_station_areas.Copy()
+
+	var/list/area/grand_list_of_areas = get_all_existing_areas_of_types(specific_areas)
+	. = list()
+	for(var/thing in shuffle(grand_list_of_areas))
+		var/area/A = thing
+		if(A.forbid_events)
+			continue
+		if(!(A.z in get_location_z_levels()))
+			continue
+		if(!ignore_occupancy && is_area_occupied(A))
+			continue // Occupied.
+		. += A
+
+// Starts the event.
+/datum/event2/event/proc/execute()
+	time_started = world.time
+
+	if(announce_delay_lower_bound)
+		announce_delay = rand(announce_delay_lower_bound, announce_delay_upper_bound ? announce_delay_upper_bound : announce_delay_lower_bound)
+		time_to_announce = world.time + announce_delay
+
+	if(start_delay_lower_bound)
+		start_delay = rand(start_delay_lower_bound, start_delay_upper_bound ? start_delay_upper_bound : start_delay_lower_bound)
+		time_to_start = world.time + start_delay
+
+	if(length_lower_bound)
+		var/starting_point = time_to_start ? time_to_start : world.time
+		length = rand(length_lower_bound, length_upper_bound ? length_upper_bound : length_lower_bound)
+		time_to_end = starting_point + length
+
+	set_up()
+
+// Called at the very end of the event's lifecycle, or when aborted.
+// Don't override this, use `end()` for cleanup instead.
+/datum/event2/event/proc/finish()
+	finished = TRUE
+	time_finished = world.time
+
+// Called by admins wanting to stop an event immediately.
+/datum/event2/event/proc/abort()
+	if(!announced)
+		announce()
+	if(!ended) // `end()` generally has cleanup procs, so call that.
+		end()
+	finish()
+
+// Called by the GM processer.
+/datum/event2/event/process()
+	// Handle announcement track.
+	if(!announced && should_announce())
+		announced = TRUE
+		announce()

```

Comment:
I'd prefer to have them be outside or be wrapped with something like `do_announce()` since setting `announced = TRUE` in `announce()` would require all events to call `..()` and it can be rather easy to forget to do that and not realize its broken.